queries/typescript/scopeTypes.scm

```diff
@@ -0,0 +1,57 @@
+(object) @map
+
+(arrow_function) @anonymousFunction
+(function) @anonymousFunction
+
+; We could also have used an alternation here, but the above style is clean and
+; seems to be more commonly used.
+; [
+;     (arrow_function)
+;     (function)
+; ] @anonymousFunction 
+
+; The following doesn't work because we can't have a `(` before the field name (`key`)
+; (
+;     (object
+;         (pair
+;             (key: (_) @key
+;             .
+;             ":") @key.removalRange
+;         ) @key.searchScope
```

Comment:
One thing to note for all our definitions of `key` is that the trailing `,` on the `pair` is not included in the `searchScope`.  We could probably handle that one fairly easily, but I do wonder if we risk people forgetting to do this.  As an alternative, if we had the `searchScope` instead point to the `pair` scope type, then we could potentially get that delimiter information for free

Reply:
What is `@key.searchScope` meant to do here? Wouldn’t we need to manually match the “,” token with a match, and bind it to, e.g, `@key.removalRange`?

Reply:
`key.searchScope` enables us to say "take key" from anywhere within the pair that contains the key, instead of the default searching behaviour which only matches if you're inside the scope itself

Reply:
So I tried the obvious thing for `,`:

```
(
    (object
        (
            (pair
                key: (_) @key
                ":" @key.delimiter.trailing
            )
            ","?
        ) @key.searchScope
    ) @key.iterationScope
)
```

And the problem is that for sibling matches, it appears to only return the range of the first sibling, rather than the range of the two siblings.  @Will-Sommers / @wenkokke have you seen this kind of thing?  Do you know if we can get the full range from the web-tree-sitter api?

Here's the output on simplified code.

Query:

```
(((pair) ",") @test1)
(((pair) (pair)) @test2)
(("," (pair)) @test3)
```

Code:

```typescript
foo = { bar: "baz", bongo: "bazman" };
```

Output:

```
    pattern:  0, capture: 0 - test1, start: (0, 8), end: (0, 18), text: `bar: "baz"`
    pattern:  1, capture: 1 - test2, start: (0, 8), end: (0, 18), text: `bar: "baz"`
    pattern:  2, capture: 2 - test3, start: (0, 18), end: (0, 19), text: `,`
```

Reply:
Heyo — I think what we're looking for here is the `[ ]` alternator syntax. This will allow for a match of the first case or the second case.

```
[
        ((pair) ",")
        (pair)
] @pair
```
yields separate match groups for each kv pair. If we add a `*` at the end, it will be grouped together. If we add a tag at the end of the `,` it will be included in the group as a tag. It would normally be excluded as it isn't a named node.

Is this what we want?

```
foo = {
  bar: "baz",
  bongo: "bazman",
  yar: "yaz"
};


(object
    [
        ((pair) "," @delim)
        (pair)
    ]* @pair
)

ts-test.ts
  pattern: 2
    capture: 2 - pair, start: (1, 2), end: (1, 12), text: `bar: "baz"`
    capture: 1 - delim, start: (1, 12), end: (1, 13), text: `,`
    capture: 2 - pair, start: (2, 2), end: (2, 17), text: `bongo: "bazman"`
    capture: 1 - delim, start: (2, 17), end: (2, 18), text: `,`
    capture: 2 - pair, start: (3, 2), end: (3, 12), text: `yar: "yaz"`
```

Reply:
I'm going to open a ticket on the main project to check this, alternative in the breach is to check for the presence of a named `delim` node after a pair/key and add it to the removal list. At least it is in sequential order.

Reply:
Interesting.  Fwiw my goal for this specific experiment isn't removal range, but search range.  Ie I want the following to work:

```typescript
{foo: "bar", baz: "bongo"}
```

If your cursor is right after the `,` I'd like to be able to say "take key" and get `foo`.

I guess we could use the same workaround we were using for `removalRange`, where we allow a `.end` tag.  So we could say:

```
(
    (object
         (pair
             key: (_) @key
             ":" @key.delimiter.trailing
         ) @key.searchScope
         ","? @key.searchScope.end
    ) @key.iterationScope
)
```

Reply:
Got it, got it.